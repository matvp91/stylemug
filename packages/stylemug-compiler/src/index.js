export { compileSchema } from './compile';

export { extractCss, save } from './extractor';
