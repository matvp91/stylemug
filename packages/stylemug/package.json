{
  "name": "stylemug",
  "version": "1.0.14",
  "main": "dist/index.js",
  "homepage": "https://github.com/matvp91/stylemug",
  "license": "MIT",
  "author": "Matthias Van Parijs",
  "funding": {
    "type": "github",
    "url": "https://github.com/sponsors/matvp91"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/matvp91/stylemug.git"
  },
  "files": [
    "dist",
    "compiler.js"
  ],
  "scripts": {
    "build": "yarn run build:prod && yarn run build:dev && yarn run build:compiler",
    "build:prod": "webpack --mode production --scope runtime",
    "build:dev": "webpack --mode development --scope runtime",
    "build:compiler": "webpack --mode production --scope compiler"
  },
  "devDependencies": {
    "babel-stylemug-plugin": "^1.0.0"
  },
  "keywords": [
    "css-in-js",
    "atomic css",
    "babel plugin",
    "css",
    "stylesheet",
    "css in js"
  ]
}
