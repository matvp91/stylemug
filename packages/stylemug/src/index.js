import warn from './warn';

export default {
  create(schema, error) {
    if (__DEV__ && error) {
      warn(error);
    }

    const resolver = (...classNames) => {
      const maps = [{}];

      const len = classNames.length;
      for (let i = 0; i < len; i++) {
        const className = classNames[i];
        if (!className) {
          continue;
        }
        if (typeof className === 'object') {
          maps.push(className);
        }
        if (typeof className === 'string') {
          maps.push(schema[className]);
        }
      }

      return Object.values(Object.assign.apply(null, maps)).join(' ');
    };

    for (let key in schema) {
      resolver[key] = schema[key];
    }

    return resolver;
  },
};
