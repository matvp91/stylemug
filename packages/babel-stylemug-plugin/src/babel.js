import jsx from '@babel/plugin-syntax-jsx';
import evaluateSimple from 'babel-helper-evaluate-path';
import { compileSchema } from 'stylemug-compiler';

export function babelPlugin(babel) {
  const t = babel.types;

  function wrapReports(path, reports) {
    if (!Array.isArray(reports)) {
      reports = [reports];
    }
    if (!reports.length) {
      return;
    }

    const node = t.cloneDeep(path.node);
    const objectChilds = [
      t.objectProperty(
        t.identifier('sourceLinesRange'),
        t.stringLiteral(
          'In lines ' + node.loc.start.line + ' to ' + node.loc.end.line
        )
      ),
      t.objectProperty(
        t.identifier('messages'),
        t.arrayExpression(
          reports.map((report) => t.stringLiteral(report.message))
        )
      ),
    ];
    if (path.hub.file.opts.filename) {
      objectChilds.push(
        t.objectProperty(
          t.identifier('fileName'),
          t.stringLiteral(path.hub.file.opts.filename || 'unknown')
        )
      );
    }

    node.arguments[1] = t.objectExpression(objectChilds);
    path.replaceWith(node);
  }

  return {
    name: 'stylemug/babel',
    inherits: jsx,
    visitor: {
      ImportDeclaration(path) {
        const specifier = path.get('specifiers').find((specifier) => {
          return (
            specifier.isImportDefaultSpecifier() &&
            specifier.parent.source.value === 'stylemug'
          );
        });

        if (!specifier) {
          return;
        }

        const references = path.scope.getBinding(specifier.node.local.name)
          .referencePaths;

        // Collect each stylemug.create({}) reference.
        references.forEach((reference) => {
          const local = reference.parentPath.parentPath;

          const sheet = evaluateSimple(local.get('arguments')[0]);
          if (!sheet.confident) {
            wrapReports(local, {
              message:
                'Failed to evaluate the following stylesheet: \n\n' +
                local.toString() +
                '\n\n' +
                'Make sure your stylesheet is statically defined.',
            });
            return;
          }

          const { result, reports } = compileSchema(sheet.value);
          wrapReports(local, reports);

          const nextLocal = t.cloneDeep(local.node);
          nextLocal.arguments[0] = t.objectExpression(
            Object.entries(result).map(([name, rules]) => {
              return t.objectProperty(
                t.identifier(name),
                t.objectExpression(
                  Object.entries(rules).map(([hash, rule]) => {
                    return t.objectProperty(
                      t.identifier(rule.keyId),
                      t.stringLiteral(hash)
                    );
                  })
                )
              );
            })
          );

          local.replaceWith(nextLocal);
        });
      },
    },
  };
}
