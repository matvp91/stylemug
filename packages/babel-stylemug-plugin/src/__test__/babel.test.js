import { transform } from '@babel/core';
import { compileSchema } from 'stylemug-compiler';
import { babelPlugin as plugin } from '../babel';

jest.mock('stylemug-compiler', () => ({
  compileSchema: jest.fn(),
}));

describe('babel plugin', () => {
  beforeEach(() => {
    compileSchema.mockImplementationOnce(() => ({
      result: {
        className: {
          hash: {
            keyId: 'id',
          },
        },
      },
      reports: [],
    }));
  });

  it('should replace create argument', () => {
    const example = `
      import stylemug from 'stylemug';

      const styles = stylemug.create({
        default: {
          color: 'red',
        },
      });
    `;

    const { code } = transform(example, {
      plugins: [plugin],
    });

    expect(code).toMatchSnapshot();
  });

  it('should error when failed to resolve', () => {
    const example = `
      import stylemug from 'stylemug';
      import { selector } from './mock';

      const styles = stylemug.create({
        [selector]: {
          color: 'red',
        },
      });
    `;

    const { code } = transform(example, {
      plugins: [plugin],
    });

    expect(code).toMatchSnapshot();
  });

  it('should replace create argument multiple times in the same file', () => {
    const example = `
      import stylemug from 'stylemug';

      const styles1 = stylemug.create({
        default: {
          color: 'red',
        },
      });

      const styles2 = stylemug.create({
        default: {
          color: 'blue',
        },
      });
    `;

    const { code } = transform(example, {
      plugins: [plugin],
    });

    expect(code).toMatchSnapshot();
  });
});

describe('babel plugin - reports', () => {
  it('should replace create argument', () => {
    compileSchema.mockImplementationOnce(() => ({
      result: {},
      reports: [{ message: 'mockError1' }, { message: 'mockError2' }],
    }));

    const example = `
      import stylemug from 'stylemug';

      const styles = stylemug.create({});
    `;

    const { code } = transform(example, {
      plugins: [plugin],
    });

    expect(code).toMatchSnapshot();
  });
});
