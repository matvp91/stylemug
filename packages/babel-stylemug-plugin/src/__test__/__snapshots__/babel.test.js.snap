// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`babel plugin - reports should replace create argument 1`] = `
"\\"use strict\\";

var _stylemug = _interopRequireDefault(require(\\"stylemug\\"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { \\"default\\": obj }; }

var styles = _stylemug[\\"default\\"].create({}, {
  sourceLinesRange: \\"In lines 4 to 4\\",
  messages: [\\"mockError1\\", \\"mockError2\\"]
});"
`;

exports[`babel plugin should error when failed to resolve 1`] = `
"\\"use strict\\";

var _stylemug = _interopRequireDefault(require(\\"stylemug\\"));

var _mock = require(\\"./mock\\");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { \\"default\\": obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var styles = _stylemug[\\"default\\"].create(_defineProperty({}, _mock.selector, {
  color: 'red'
}), {
  sourceLinesRange: \\"In lines 5 to 9\\",
  messages: [\\"Failed to evaluate the following stylesheet: \\\\n\\\\nstylemug.create({\\\\n  [selector]: {\\\\n    color: 'red'\\\\n  }\\\\n})\\\\n\\\\nMake sure your stylesheet is statically defined.\\"]
});"
`;

exports[`babel plugin should replace create argument 1`] = `
"\\"use strict\\";

var _stylemug = _interopRequireDefault(require(\\"stylemug\\"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { \\"default\\": obj }; }

var styles = _stylemug[\\"default\\"].create({
  className: {
    id: \\"hash\\"
  }
});"
`;

exports[`babel plugin should replace create argument multiple times in the same file 1`] = `
"\\"use strict\\";

var _stylemug = _interopRequireDefault(require(\\"stylemug\\"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { \\"default\\": obj }; }

var styles1 = _stylemug[\\"default\\"].create({
  className: {
    id: \\"hash\\"
  }
});

var styles2 = _stylemug[\\"default\\"].create({
  className: {
    id: \\"hash\\"
  }
});"
`;
